File: Queue.cpp
#include "Queue.h"
#include <iostream>
#include <fstream>
#include <vector>
#include <stdexcept>

using namespace std;

Queue::Queue() {
    front = nullptr;
    rear = nullptr;
    size = 0;
}

Queue::~Queue() {
    QCLEAR();
}

void Queue::QCLEAR() {
    while (front != nullptr) {
        QNode* temp = front;
        front = front->next;
        delete temp;
    }
    rear = nullptr;
    size = 0;
    cout << "Очередь очищена." << endl;
}

void Queue::QPUSH(const string& value) {
    QNode* newNode = new QNode{ value, nullptr };

    if (rear == nullptr) {
        // Очередь пуста - новый элемент становится началом и концом
        front = newNode;
        rear = newNode;
    }
    else {
        // Добавляем в конец очереди
        rear->next = newNode;
        rear = newNode;
    }

    size++;
    //cout << "Элемент \"" << value << "\" добавлен в очередь" << endl;
}

string Queue::QPOP() {
    if (front == nullptr) {
        throw runtime_error("Ошибка: очередь пуста!");
    }

    QNode* temp = front;
    string value = temp->data;
    front = front->next;

    // Если очередь стала пустой, обнуляем rear
    if (front == nullptr) {
        rear = nullptr;
    }

    delete temp;
    size--;

    //cout << "Элемент \"" << value << "\" извлечён из очереди" << endl;
    return value;
}

string Queue::QFRONT() const {
    if (front == nullptr) {
        throw runtime_error("Ошибка: очередь пуста!");
    }
    return front->data;
}

void Queue::QPRINT() const {
    if (front == nullptr) {
        cout << "Очередь пуста!" << endl;
        return;
    }

    cout << "Очередь [" << size << "] (начало-> конец): ";
    QNode* current = front;
    while (current != nullptr) {
        cout << "\"" << current->data << "\"";
        if (current->next != nullptr) {
            cout << " -> ";
        }
        current = current->next;
    }
    cout << " -> NULL" << endl;
}
bool Queue::isEmpty() const {
    return size == 0;
}

File: queue.go
package main

import (
	"fmt"
)

type QNode struct {
	data string
	next *QNode
}

type Queue struct {
	front *QNode
	rear  *QNode
	size  int
}

func NewQueue() *Queue {
	return &Queue{
		front: nil,
		rear:  nil,
		size:  0,
	}
}

func (q *Queue) QCLEAR() {
	q.front = nil
	q.rear = nil
	q.size = 0
	fmt.Println("Очередь очищена.")
}

func (q *Queue) QPUSH(value string) {
	newNode := &QNode{
		data: value,
		next: nil,
	}

	if q.rear == nil {
		q.front = newNode
		q.rear = newNode
	} else {
		q.rear.next = newNode
		q.rear = newNode
	}
	q.size++
}

func (q *Queue) QPOP() (string, error) {
	if q.front == nil {
		return "", fmt.Errorf("ошибка: очередь пуста")
	}

	value := q.front.data
	q.front = q.front.next

	if q.front == nil {
		q.rear = nil
	}
	q.size--
	return value, nil
}

func (q *Queue) QFRONT() (string, error) {
	if q.front == nil {
		return "", fmt.Errorf("ошибка: очередь пуста")
	}
	return q.front.data, nil
}

func (q *Queue) QPRINT() {
	if q.front == nil {
		fmt.Println("Очередь пуста!")
		return
	}

	fmt.Printf("Очередь [%d] (начало -> конец): ", q.size)
	current := q.front
	for current != nil {
		fmt.Printf("\"%s\"", current.data)
		if current.next != nil {
			fmt.Print(" -> ")
		}
		current = current.next
	}
	fmt.Println(" -> NULL")
}

func (q *Queue) IsEmpty() bool {
	return q.size == 0
}

func (q *Queue) Size() int {
	return q.size
}

// Для сериализации
func (q *Queue) GetAll() []string {
	result := make([]string, 0, q.size)
	current := q.front
	for current != nil {
		result = append(result, current.data)
		current = current.next
	}
	return result
}

func (q *Queue) SetAll(items []string) {
	q.QCLEAR()
	for _, item := range items {
		q.QPUSH(item)
	}
}